File: PublicIP.fs
module PublicIP

open FSharpTools
open FSharpHttpRequest
open AsyncResult

let get () = 
    let extractIp = 
        String.subStringBetweenStrs "Address: " "<"
        >> (Option.defaultValue "")
        >> Async.toAsync

    Request.getString { Request.defaultSettings with Url = "http://checkip.dyndns.org" }
    |>> extractIp


    // TODO sometimes 502 bad gateway!
    

File: PublicIP.cs
using System;
using System.Net.Http;
using System.Threading.Tasks;
using CsTools.Async;
using CsTools.Extensions;
using CsTools.HttpRequest;

using static CsTools.Core;
using static CsTools.HttpRequest.Core;

static class PublicIP
{
    public static Task<string> Get()
        => RepeatOnException(
            () => RequestIP()
                    .Select(ExtractIP),
            4, 
            TimeSpan.FromSeconds(5));

    static string ExtractIP(string ipStr)
        => ipStr.StringBetween("Address: ", "<");

    static Task<string> RequestIP()
        => Request.GetStringAsync(DefaultSettings with
        {
            Method = HttpMethod.Get,
            BaseUrl = "http://checkip.dyndns.org"
        });
}
